From: Hannes Reinecke <hare@suse.de>
Date: Mon, 12 Sep 2016 12:35:42 +0200
Subject: fc_rport: update fcp_parm after calling prli provider
Patch-Mainline: submitted linux-scsi 2016/09/12
References: bsc#998385

fc_rport_prli_resp() checks the PRLI response of the remote port.
On the initiator side, fc_passive_prov[FC_TYPE_FCP] may be non-NULL
if tcm_fc is loaded (maybe this system is also acting as target on a
different port). ft_prli() returns 0 in this case, causing the
FCP_SPPF_TARG_FCN to be cleared.
But we need to re-evaluate fcp_parm to set the FC roles in the
rport correctly.

Reported-by: Martin Wilck <martin.wilck@suse.com>
Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 drivers/scsi/libfc/fc_rport.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/scsi/libfc/fc_rport.c b/drivers/scsi/libfc/fc_rport.c
index d070cbf..e893c71 100644
--- a/drivers/scsi/libfc/fc_rport.c
+++ b/drivers/scsi/libfc/fc_rport.c
@@ -1189,7 +1189,10 @@ static void fc_rport_prli_resp(struct fc_seq *sp, struct fc_frame *fp,
 		/*
 		 * Check if the image pair could be established
 		 */
-		if (rdata->spp_type != FC_TYPE_FCP ||
+		if (!resp_code)
+			/* ->prli might have changed parameters */
+			fcp_parm = ntohl(pp->spp.spp_params);
+		else if (rdata->spp_type != FC_TYPE_FCP ||
 		    resp_code != FC_SPP_RESP_ACK ||
 		    !(temp_spp.spp_flags & FC_SPP_EST_IMG_PAIR)) {
 			/*
-- 
1.8.5.6

